### ISSUE TITLE
[BUG] Improper chapter generation for mkv

### ISSUE THREAD

**Issue by hazy-kun** [stratuma/Crunchyroll-Downloader-v4.0#22]
**Describe the bug**
Chapters file is improperly generated. endtime is not supported nor parsed from mkv container

**To Reproduce**
Download an episode that have chapters end open mpv

**Expected behavior**
You need to convert the end time in a dummy chapter that have the end time of a real chapter as its start time in order to properly view and use the chapters in mpv or other player

**Desktop (please complete the following information):**
 - OS: Windows
 - Browser all
 - Version 1.2.0

**Additional context**
The problem is not properly a bug but incorrect chapters generation make it unusable

**Comment by stratuma** [comment]
```
export async function createChapterFile(rawchapters: VideoMetadata, dir: string) {
    const filepath = path.join(dir, 'chapters.txt')

    var chapters: string[] = []

    chapters.push(';FFMETADATA1')

    if (rawchapters.intro && rawchapters.intro.type && rawchapters.intro.start && rawchapters.intro.end) {
        chapters.push('[CHAPTER]')
        chapters.push('TIMEBASE=1/1000')
        chapters.push(`START=${formatTimeFFMPEG(rawchapters.intro.start)}`)
        chapters.push(`END=${formatTimeFFMPEG(rawchapters.intro.end) - 1}`)
        chapters.push('title=Intro')
    }

    if (rawchapters.credits && rawchapters.credits.type && rawchapters.credits.start && rawchapters.credits.end) {
        chapters.push('[CHAPTER]')
        chapters.push('TIMEBASE=1/1000')
        chapters.push(`START=${formatTimeFFMPEG(rawchapters.credits.start)}`)
        chapters.push(`END=${formatTimeFFMPEG(rawchapters.credits.end) - 1}`)
        chapters.push('title=Credits')
    }

    if (rawchapters.preview && rawchapters.preview.type && rawchapters.preview.start && rawchapters.preview.end) {
        chapters.push('[CHAPTER]')
        chapters.push('TIMEBASE=1/1000')
        chapters.push(`START=${formatTimeFFMPEG(rawchapters.preview.start)}`)
        chapters.push(`END=${formatTimeFFMPEG(rawchapters.preview.end) - 1}`)
        chapters.push('title=Preview')
    }

    if (rawchapters.recap && rawchapters.recap.type && rawchapters.recap.start && rawchapters.recap.end) {
        chapters.push('[CHAPTER]')
        chapters.push('TIMEBASE=1/1000')
        chapters.push(`START=${formatTimeFFMPEG(rawchapters.recap.start)}`)
        chapters.push(`END=${formatTimeFFMPEG(rawchapters.recap.end) - 1}`)
        chapters.push('title=Recap')
    }

    try {
        await fs.writeFile(filepath, chapters.join('\r\n'), 'utf-8')
    } catch (e) {
        console.error(`Error occurred during chapter writing:`, e)
        server.logger.log({
            level: 'error',
            message: `Error occurred during chapter writing:`,
            error: e,
            timestamp: new Date().toISOString(),
            section: 'crunchyrollWritingProcessChapter'
        })
    }

    return filepath
}
```

What I know this is the right way to do it, it could still be wrong, but I add the endtime.

**Comment by stratuma** [comment]
https://www.terrybutler.co.uk/2021/08/01/how-to-add-chapters-to-video-using-ffmpeg/

**Comment by hazy-kun** [comment]
![Screenshot (1)](https://github.com/stratuma/Crunchyroll-Downloader-v4.0/assets/145157126/342ff24f-7749-494c-84f3-11a24de1e61f)
As you can see there are three chpaters, one without name, one named intro but finish with the start of credits chapter, and credits.
In case of non parsed by CR after credit scenes the credits chapter will end improperly at the end of the video not at the end of the ED

**Comment by stratuma** [comment]
The one without name is maybe the END tag, but I still don't know how to solve it, since its the right way provided by ffmpeg

**Comment by hazy-kun** [comment]
I attach an example, 01 is your chapters file for GG1U2NNZ7, 02 is the version corrected

01
`<?xml version="1.0"?>

<Chapters>
  <EditionEntry>
    <EditionFlagDefault>1</EditionFlagDefault>
    <ChapterAtom>
      <ChapterUID>1</ChapterUID>
      <ChapterTimeStart>00:02:52.000000000</ChapterTimeStart>
      <ChapterTimeEnd>00:04:18.999000000</ChapterTimeEnd>
      <ChapterDisplay>
        <ChapterString>Intro</ChapterString>
        <ChapterLanguage>und</ChapterLanguage>
      </ChapterDisplay>
    </ChapterAtom>
    <ChapterAtom>
      <ChapterUID>2</ChapterUID>
      <ChapterTimeStart>00:22:03.000000000</ChapterTimeStart>
      <ChapterTimeEnd>00:23:32.999000000</ChapterTimeEnd>
      <ChapterDisplay>
        <ChapterString>Credits</ChapterString>
        <ChapterLanguage>und</ChapterLanguage>
      </ChapterDisplay>
    </ChapterAtom>
  </EditionEntry>
</Chapters>
`

02
`<?xml version="1.0"?>

<Chapters>
  <EditionEntry>
    <EditionUID>2370877332122925185</EditionUID>
    <ChapterAtom>
      <ChapterTimeStart>00:02:53.000000000</ChapterTimeStart>
      <ChapterUID>3756700515759439282</ChapterUID>
      <ChapterDisplay>
        <ChapterString>Intro</ChapterString>
        <ChapterLanguage>und</ChapterLanguage>
        <ChapLanguageIETF>und</ChapLanguageIETF>
      </ChapterDisplay>
    </ChapterAtom>
    <ChapterAtom>
      <ChapterTimeStart>00:04:35.000000000</ChapterTimeStart>
      <ChapterUID>1661436401791150151</ChapterUID>
      <ChapterDisplay>
        <ChapterString>Chapter</ChapterString>
        <ChapterLanguage>und</ChapterLanguage>
        <ChapLanguageIETF>und</ChapLanguageIETF>
      </ChapterDisplay>
    </ChapterAtom>
    <ChapterAtom>
      <ChapterTimeStart>00:22:05.000000000</ChapterTimeStart>
      <ChapterUID>4994504243197762647</ChapterUID>
      <ChapterDisplay>
        <ChapterString>Credits</ChapterString>
        <ChapterLanguage>und</ChapterLanguage>
        <ChapLanguageIETF>und</ChapLanguageIETF>
      </ChapterDisplay>
    </ChapterAtom>
    <ChapterAtom>
      <ChapterTimeStart>00:23:35.000000000</ChapterTimeStart>
      <ChapterUID>1801854749204984734</ChapterUID>
      <ChapterDisplay>
        <ChapterString>Chapter</ChapterString>
        <ChapterLanguage>und</ChapterLanguage>
        <ChapLanguageIETF>und</ChapLanguageIETF>
      </ChapterDisplay>
    </ChapterAtom>
    <ChapterAtom>
      <ChapterTimeStart>00:23:35.000000000</ChapterTimeStart>
      <ChapterUID>15454288063513380056</ChapterUID>
      <ChapterDisplay>
        <ChapterString>Preview</ChapterString>
        <ChapterLanguage>und</ChapterLanguage>
        <ChapLanguageIETF>und</ChapLanguageIETF>
      </ChapterDisplay>
    </ChapterAtom>
  </EditionEntry>
</Chapters>
`

try to mux to see the differences

**Comment by hazy-kun** [comment]
YOUR
```
<?xml version="1.0"?>
<!-- <!DOCTYPE Chapters SYSTEM "matroskachapters.dtd"> -->
<Chapters>
  <EditionEntry>
    <EditionFlagDefault>1</EditionFlagDefault>
    <ChapterAtom>
      <ChapterUID>1</ChapterUID>
      <ChapterTimeStart>00:02:52.000000000</ChapterTimeStart>
      <ChapterTimeEnd>00:04:18.999000000</ChapterTimeEnd>
      <ChapterDisplay>
        <ChapterString>Intro</ChapterString>
        <ChapterLanguage>und</ChapterLanguage>
      </ChapterDisplay>
    </ChapterAtom>
    <ChapterAtom>
      <ChapterUID>2</ChapterUID>
      <ChapterTimeStart>00:22:03.000000000</ChapterTimeStart>
      <ChapterTimeEnd>00:23:32.999000000</ChapterTimeEnd>
      <ChapterDisplay>
        <ChapterString>Credits</ChapterString>
        <ChapterLanguage>und</ChapterLanguage>
      </ChapterDisplay>
    </ChapterAtom>
  </EditionEntry>
</Chapters>
```

CHAPTERS MKV USABLE
```
<?xml version="1.0"?>
<!-- <!DOCTYPE Chapters SYSTEM "matroskachapters.dtd"> -->
<Chapters>
  <EditionEntry>
    <EditionUID>2370877332122925185</EditionUID>
    <ChapterAtom>
      <ChapterTimeStart>00:02:53.000000000</ChapterTimeStart>
      <ChapterUID>3756700515759439282</ChapterUID>
      <ChapterDisplay>
        <ChapterString>Intro</ChapterString>
        <ChapterLanguage>und</ChapterLanguage>
        <ChapLanguageIETF>und</ChapLanguageIETF>
      </ChapterDisplay>
    </ChapterAtom>
    <ChapterAtom>
      <ChapterTimeStart>00:04:35.000000000</ChapterTimeStart>
      <ChapterUID>1661436401791150151</ChapterUID>
      <ChapterDisplay>
        <ChapterString>Chapter</ChapterString>
        <ChapterLanguage>und</ChapterLanguage>
        <ChapLanguageIETF>und</ChapLanguageIETF>
      </ChapterDisplay>
    </ChapterAtom>
    <ChapterAtom>
      <ChapterTimeStart>00:22:05.000000000</ChapterTimeStart>
      <ChapterUID>4994504243197762647</ChapterUID>
      <ChapterDisplay>
        <ChapterString>Credits</ChapterString>
        <ChapterLanguage>und</ChapterLanguage>
        <ChapLanguageIETF>und</ChapLanguageIETF>
      </ChapterDisplay>
    </ChapterAtom>
    <ChapterAtom>
      <ChapterTimeStart>00:23:35.000000000</ChapterTimeStart>
      <ChapterUID>1801854749204984734</ChapterUID>
      <ChapterDisplay>
        <ChapterString>Chapter</ChapterString>
        <ChapterLanguage>und</ChapterLanguage>
        <ChapLanguageIETF>und</ChapLanguageIETF>
      </ChapterDisplay>
    </ChapterAtom>
    <ChapterAtom>
      <ChapterTimeStart>00:23:35.000000000</ChapterTimeStart>
      <ChapterUID>15454288063513380056</ChapterUID>
      <ChapterDisplay>
        <ChapterString>Preview</ChapterString>
        <ChapterLanguage>und</ChapterLanguage>
        <ChapLanguageIETF>und</ChapLanguageIETF>
      </ChapterDisplay>
    </ChapterAtom>
  </EditionEntry>
</Chapters>
```

Same response for GG1U2NNZ7, try muxing and see the differences

**Comment by stratuma** [comment]
Your version is missing the end tags.
```
<ChapterAtom>
      <ChapterTimeStart>00:04:35.000000000</ChapterTimeStart>
      <ChapterUID>1661436401791150151</ChapterUID>
      <ChapterDisplay>
        <ChapterString>Chapter</ChapterString>
        <ChapterLanguage>und</ChapterLanguage>
        <ChapLanguageIETF>und</ChapLanguageIETF>
      </ChapterDisplay>
    </ChapterAtom>
```

And what is this? Crunchyroll doesn't have a metadata called Chapter

**Comment by hazy-kun** [comment]
> Your version is missing the end tags.
> 
> ```
> <ChapterAtom>
>       <ChapterTimeStart>00:04:35.000000000</ChapterTimeStart>
>       <ChapterUID>1661436401791150151</ChapterUID>
>       <ChapterDisplay>
>         <ChapterString>Chapter</ChapterString>
>         <ChapterLanguage>und</ChapterLanguage>
>         <ChapLanguageIETF>und</ChapLanguageIETF>
>       </ChapterDisplay>
>     </ChapterAtom>
> ```
> 
> And what is this? Crunchyroll doesn't have a metadata called Chapter

Yes as said before I've converted end tangs in dummy schapter start tags 'cause mkv doesn't support end tags

Please read the issues before aswer, the end tag in your file is not parsed by the player

**Comment by stratuma** [comment]
You are right, sorry, my error

**Comment by stratuma** [comment]
```
async function createChapter(chapter: VideoMetadataSingle, format: string, name: string) {
    var chapters: string[] = []
    if (format === 'mkv') {
        chapters.push('[CHAPTER]')
        chapters.push('TIMEBASE=1/1000')
        chapters.push(`START=${formatTimeFFMPEG(chapter.start)}`)
        chapters.push(`END=${formatTimeFFMPEG(chapter.start)}`)
        chapters.push(`title=${name}`)
        chapters.push(``)
        chapters.push('[CHAPTER]')
        chapters.push('TIMEBASE=1/1000')
        chapters.push(`START=${formatTimeFFMPEG(chapter.end)}`)
        chapters.push(`END=${formatTimeFFMPEG(chapter.end)}`)
        chapters.push(`title=Chapter`)
        chapters.push(``)
    } else {
        chapters.push('[CHAPTER]')
        chapters.push('TIMEBASE=1/1000')
        chapters.push(`START=${formatTimeFFMPEG(chapter.start)}`)
        chapters.push(`END=${formatTimeFFMPEG(chapter.end)}`)
        chapters.push(`title=${name}`)
        chapters.push(``)
    }

    return chapters
}

export async function createChapterFile(rawchapters: VideoMetadata, dir: string, format: string, mediaid: string) {
    const filepath = path.join(dir, 'chapters.txt')

    var chapters: string[] = []

    chapters.push(';FFMETADATA1')
    chapters.push(`title=${mediaid}`)
    chapters.push(`artist=Crunchyroll`)
    chapters.push(``)

    if (rawchapters.intro && rawchapters.intro.type && rawchapters.intro.start && rawchapters.intro.end) {
        chapters.push(...await createChapter(rawchapters.intro, format, 'Intro'));
        console.log(chapters)
    }

    if (rawchapters.credits && rawchapters.credits.type && rawchapters.credits.start && rawchapters.credits.end) {
        chapters.push(...await createChapter(rawchapters.credits, format, 'Credits'));
        console.log(chapters)
    }

    if (rawchapters.preview && rawchapters.preview.type && rawchapters.preview.start && rawchapters.preview.end) {
        chapters.push(...await createChapter(rawchapters.preview, format, 'Preview'));
        console.log(chapters)
    }

    if (rawchapters.recap && rawchapters.recap.type && rawchapters.recap.start && rawchapters.recap.end) {
        chapters.push(...await createChapter(rawchapters.recap, format, 'Recap'));
        console.log(chapters)
    }

    try {
        await fs.writeFile(filepath, chapters.join('\r\n'), 'utf-8')
    } catch (e) {
        console.error(`Error occurred during chapter writing:`, e)
        server.logger.log({
            level: 'error',
            message: `Error occurred during chapter writing:`,
            error: e,
            timestamp: new Date().toISOString(),
            section: 'crunchyrollWritingProcessChapter'
        })
    }

    return filepath
}
```

I let the END tag in because without it the chapter name somehow doesn't appear, but like this it creates dummy end chapters

**Comment by stratuma** [comment]
![grafik](https://github.com/stratuma/Crunchyroll-Downloader-v4.0/assets/166541445/b0357214-39f6-4faf-a84a-97d529140d52)

**Comment by stratuma** [comment]
Fixed in v1.2.1
